dj/typing.py

```diff
@@ -46,27 +50,133 @@ class SQLADialect(Protocol):  # pylint: disable=too-few-public-methods
+
+            return ColumnType(attr)
+        except ColumnTypeError:
+            return type.__getattribute__(cls, attr)
+
+    def __getitem__(cls, key) -> "ColumnType":
+        return ColumnType(key)
+
+
+# pylint: disable=C0301
+class ColumnType(str, metaclass=ColumnTypeMeta):
     """
     Types for columns.
 
     These represent the values from the ``python_type`` attribute in SQLAlchemy columns.
+
+    NOTE: `ColumnType` is just a special string type and can be used anywhere a string would be
+
+        >>> ColumnType('list[INT]')
+        'LIST[INT]'
+
+        >>> ColumnType['INT']
+        'INT'
+
+        >>> ColumnType.List[ColumnType.Int]
+        'LIST[INT]'
+
+        >>> ColumnType.List[ColumnType.Int].args[0]
+        'INT'
+
+        >>> ColumnType.dict[ColumnType.INT, ColumnType.list[ColumnType.dict[ColumnType.INT, ColumnType.list[ColumnType.STR]]]]
+        'DICT[INT, LIST[DICT[INT, LIST[STR]]]]'
+
+        >>> ColumnType.Dict[ColumnType.str, ColumnType.list[ColumnType.int]].args[1].args[0]
+        'INT'
     """
 
-    NULL = "NULL"
-    BYTES = "BYTES"
-    STR = "STR"
-    FLOAT = "FLOAT"
-    INT = "INT"
-    DECIMAL = "DECIMAL"
-    BOOL = "BOOL"
-    DATETIME = "DATETIME"
-    DATE = "DATE"
-    TIME = "TIME"
-    TIMEDELTA = "TIMEDELTA"
-    LIST = "LIST"
-    DICT = "DICT"
-    WILDCARD = "*"
+    # pylint: enable=C0301
+    args = None
+
+    def __new__(cls, type_: str):
+
+        if isinstance(type_, ColumnType):
+            return type_
+        type_ = type_.upper().strip()
+        if type_ in COMPLEX_TYPES or type_ in PRIMITIVE_TYPES:
+            validated = type_
+        else:
+            validated = cls._validate_type(type_)
+
+        obj = str.__new__(cls, validated)
+        return obj
+
+    def __getitem__(self, keys) -> "ColumnType":
+        if not isinstance(keys, (list, tuple)):
+            keys = (keys,)
+        keys = tuple(keys)
+        if self not in COMPLEX_TYPES:
+            raise ColumnTypeError(f"The type {self} is not a complex type.")
+        if len(keys) != COMPLEX_TYPES[self]:
+            raise ColumnTypeError(
+                f"{self} expects {COMPLEX_TYPES[self]} inner type(s) but got {len(keys)}.",
+            )
+        args = tuple(ColumnType(key) for key in keys)
+        obj = str.__new__(
+            self.__class__,
+            self + "[" + ", ".join(args) + "]",
+        )
+        obj.args = args
+        return obj
+
+    @classmethod
+    def _validate_type(cls, type_: str):
+        test = TYPE_PATTERN.match(type_)
+        if test is not None:
+            outer = test.group("outer")
+            inner = test.group("inner")
+            if outer not in COMPLEX_TYPES:
+                raise ColumnTypeError(f"{outer} is not a KNOWN complex type.")
+            inners = inner.split(",")
+            return ColumnType(outer)[inners]
+        if type_ not in PRIMITIVE_TYPES:
+            raise ColumnTypeError(f"Expected a primitive type but got {type_}.")
```

Comment:
Should this error say that it's specifically expecting a primitive type? Seems it could also be from a typo while defining a complex type.
```py
>>> ColumnType('LIST[STR')
ColumnTypeError: Expected a primitive type but got LIST[STR.
```
```py
>>> ColumnType('LIST[STR]')
'LIST[STR]'
```

Reply:
changed it to a more generic exception message